pkg/features/feature_manager.go

```diff
@@ -0,0 +1,265 @@
+}
+
+// This method returns the Feature Manager Struct. Should not be modified after intialization.
+func GetFeatureManager() *FeatureManager {
+	return &RisingWaveFeatureManager
+}
+
+// This is a helper functions that adds a feature to the featureManager, will be used on init to init and add all
+// features to the featureManager.
+func (m *FeatureManager) addFeature(feature *Feature) {
+	m.featureMap[feature.Name] = feature.DeepCopy()
+}
+
+// This function returns true if the feature exists in the featureManager, else returns false.
+func (m FeatureManager) IsFeatureExist(featureName FeatureName) bool {
+	_, exist := m.featureMap[featureName]
+	return exist
+}
+
+// This is a helper function that helps to set a feature to a given boolean, if feature does not exist,
+// returns an error. Used in EnableFeature and DisableFeature methods. Not visible to users.
+func (m *FeatureManager) setFeatureEnable(name FeatureName, enable bool) error {
+	// check for existence of feature in map
+	_, featureExists := m.featureMap[name]
+	if featureExists {
+		m.featureMap[name].Enabled = enable
+		return nil
+	}
+	return errors.New(fmt.Sprintf("The following feature does not exist: %s", name))
+}
+
+// This method takes in a feature name and checks if it is enabled, returns a bool, nil if it exists,
+// and a false,error if it does not exist in the featureManager.
+func (m FeatureManager) IsFeatureEnabled(name FeatureName) bool {
+	// check for existence of feature in map
+	feature, featureExists := m.featureMap[name]
+	if featureExists {
+		// if feature has been actively set/unset, return enabled, else return deafult value.
+		return feature.Enabled
+	}
+	return false
+}
+
+// This metod takes in a feature name and enables it if it exists, if it does not
+// it returns an error.
+func (m *FeatureManager) EnableFeature(name FeatureName) error {
+	// check for existence of feature in map
+	return m.setFeatureEnable(name, true)
+}
+
+// This metod takes in a feature name and disables it if it exists, if it does not
+// it returns an error.
+func (m *FeatureManager) DisableFeature(name FeatureName) error {
+	// check for existence of feature in map
+	return m.setFeatureEnable(name, false)
+}
+
+// This method returns the number of features in the featureManager.
+func (m *FeatureManager) GetNumOfFeatures() int {
+	return len(m.featureMap)
+}
+
+// This method lists all features, returns a copy of the list of feature structs.
+func (m FeatureManager) ListFeatures() []Feature {
+	var featureList = []Feature{}
+	for _, feature := range m.featureMap {
+		// make a deep copy of the feature
+		featureList = append(featureList, *feature.DeepCopy())
+	}
+	return featureList
+}
+
+// This method lists all enabled features, returns a copy of the list of feature structs.
+func (m FeatureManager) ListEnabledFeatures() []Feature {
+	var featureList = []Feature{}
+	for featureName, feature := range m.featureMap {
+		if m.IsFeatureEnabled(featureName) {
```

Comment:
It seems that we already get the `feature` object here, thus the `feature.Enabled` should be enough to be the condition.